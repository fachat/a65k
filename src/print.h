/****************************************************************************

    print output
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PRINT_H
#define PRINT_H

#include <stdio.h>

#include "parser.h"
#include "print-config.h"

#define	BUF_LEN		4096

// definitions for print output columns
typedef enum {
	PRT_LINENO		= 0,		// line number
	PRT_ADDR		= 1,		// address of line
	PRT_DATA		= 2,		// bytes generated by line
	PRT_LABEL		= 3,		// label definition
	PRT_OPERATION		= 4,		// operation
	PRT_PARAM		= 5,		// parameter
	PRT_COMMENT		= 6,		// comment
	PRT_END			= 7		// unused, defined for len
} print_col_t;

typedef struct {
	FILE			*outfile;
	const char		*filename;
	const print_config_t	*cfg;
	// output buffer
	char			buf[BUF_LEN];
	int			buflen;
	// runtime flags
	// last line was empty
	int			was_empty;
	// column widths for fixed column width
	print_col_t		current_col;
	int			col_width[PRT_END];
} printer_t;

// init print module
void print_module_init();

// initialize printer output channel
printer_t *print_init();

// general print; 
// Note: switching col_no automatically goes to this column, but
// only in ascending order. Printing with same col_no is allowed
// and simply appends.
void print(printer_t *ptr, print_col_t col_no, const char *pattern, ...);

// print a statement formatted
void print_formatted_stmt(printer_t *cfg, const statement_t *stmt);

void print_out(printer_t *prt);

void print_clr(printer_t *prt);

void print_close(printer_t *prt);

#endif

