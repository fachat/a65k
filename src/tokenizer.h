/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef TOKENIZER_H
#define TOKENIZER_H

#include <stdio.h>

#include "types.h"

// token type
typedef enum {
	T_INIT		= 'i',	// initial state
	T_BRACKET	= 'b',	// any type of bracket
	T_NAME		= 'n',	// a name, like label1 do_this
	T_TOKEN		= 't',	// a token, like ( ) . ++ == <> & 
	T_STRING	= 's',	// string
	T_LITERAL	= 'l',	// a numerical literal value (decimal, hex, ...)
	T_ERROR		= 'e',	// an error has happened
	T_END		= 'x',	// end of string reached
} tok_t;

// literal type
typedef enum {
	LIT_NONE		=' ',
	LIT_DECIMAL		='d',	// a decimal number, like 1 2 3
	LIT_OCTAL		='o',	// an octal number, like &127 
	LIT_BINARY		='b',	// a binary number, like %110101
	LIT_HEX			='h',	// a hex number, like $12dA 
	LIT_OCTAL_C		='O',	// C-type an octal number, like 0127
	LIT_HEX_C		='H',	// C-type hex number, like 0x12Ad
	LIT_CHAR		='c',	// single character
	LIT_TWOCHAR		='s',	// double character string
} littype_t;

// quote type for strings
typedef enum {
	Q_SINGLE		= '\'',
	Q_DOUBLE		= '"'
} quotetype_t;

static inline bool is_string_delim(char c) {
	return (c == Q_SINGLE || c == Q_DOUBLE);
} 


typedef enum {
	OP_NONE			= 0,
	OP_EXCL 		= '!',	
	OP_AT			= '@',
	OP_XQUOTE		= '`',
	OP_SQUOTE		= '\'',
	OP_OPEN			= '(',
	OP_CLOSE		= ')',
	OP_HASH			= '#',
	OP_BOPEN		= '[',
	OP_COPEN		= '{',
	OP_BBOPEN		= 136,	// "[["
	OP_BCLOSE		= ']',
	OP_CCLOSE		= '}',
	OP_BBCLOSE		= 137,	// "]]"
	OP_STAR			= '*',
	OP_COMMA		= ',',
	OP_PLUS			= '+',	// binary as usual, as unary modified label scope
	OP_MINUS		= '-',
	OP_DIV			= '/',
	OP_COLON		= ':',
	OP_SEMICOLON		= ';',
	OP_DOT			= '.',
	OP_LARGER		= '>',
	OP_LARGEROREQUAL	= 128,	// ">="
	OP_LESS			= '<',
	OP_LESSOREQUAL		= 129,	// "<="
	OP_ASSIGN		= '=',
	OP_ASSIGNADD		= 138,	// "+="
	OP_ASSIGNSUB		= 139,	// "-="
	OP_ASSIGNMULT		= 140,	// "*="
	OP_ASSIGNDIV		= 141,	// "/="
	OP_ASSIGNMOD		= 142,	// "%="
	OP_ASSIGNBITOR		= 143,	// "|="
	OP_ASSIGNBITAND		= 144,	// "&="
	OP_ASSIGNBITXOR		= 145,	// "^="
	OP_ASSIGNSHIFTLEFT	= 146,	// "<<="
	OP_ASSIGNSHIFTRIGHT	= 147,	// ">>="
	OP_BITXOR		= '^',
	OP_BITOR		= '|',
	OP_BITAND		= '&',	// binary as usual, as unary modifies label scope
	OP_EQUAL		= 138,	// "=="
	OP_NOTEQUAL		= 130,	// "!=", "<>", "><"
	OP_SHIFTLEFT		= 131,	// "<<"
	OP_SHIFTRIGHT		= 132,	// ">>"
	OP_LOGICAND		= 133,	// "&&"
	OP_LOGICOR		= 134,	// "||"
	OP_LOGICXOR		= 135,	// "^^"
	OP_BITINV		= '~',
	OP_MOD			= '%',
	OP_DOUBLESLASH		= 153,	// "//"
	OP_RROPEN		= 154,	// "(("	- not arithm. parsed but generated by admode parser
	OP_RRCLOSE		= 155,	// "))"	- not arithm. parsed but generated by admode parser
	// register indexing
	OP_XIND			= 148, 	// ",x"
	OP_YIND			= 149,	// ",y"
	OP_ZIND			= 150,	// ",z"
	OP_SIND			= 151,	// ",s" - stack offset
	OP_BIND			= 152, 	// ",b" - base register offset
	OP_PIND			= 153, 	// ",p" - PC offset
} op_t;

typedef struct {
	op_t		type;
	const char	*print;
	char		is_modifier;
	char		is_unary;
	char		is_binary;
	char		is_index;	// like ",x"
	char		is_assign;
} op_details_t;


static inline bool is_stmt_delim(op_t t) {
	return (t == OP_COLON);
}

static inline bool is_comment_delim(op_t t) {
	return (t == OP_SEMICOLON) || (t == OP_DOUBLESLASH);
}

static inline op_t closing_op(op_t opening_bracket) {
	switch (opening_bracket) {
	case OP_OPEN: 	return OP_CLOSE;
	case OP_BOPEN:	return OP_BCLOSE;
	case OP_COPEN:	return OP_CCLOSE;
	case OP_BBOPEN:	return OP_BBCLOSE;
	case OP_RROPEN:	return OP_RRCLOSE;
	case OP_LESS:	return OP_LARGER;
	default:	return OP_NONE;
	}
}

extern int prio_of_operator_table[];

static inline int prio_of_operator(const op_t op) {
	return prio_of_operator_table[op];
}

typedef struct {
	const char 	*line;		// line to tokenize
	int		cstyle_allowed;	// config - C-style octal and hex allowed
	tok_t		type;		// type of token
	int 		ptr;		// current pointer to token in line
	int		len;		// length of current token (total, i.e. incl. quotes for a string)
	union {
		struct {
			int		ptr;		// pointer to start of string for T_STRING_LITERAL
			int		len;		// length of string for T_STRING_LITERAL
			quotetype_t	type;		// what quote was used?
		} string;
		struct {
			long		value;		// value for decimal, octal, binary, or hex literal
			littype_t	type;		// type of literal, for T_LITERAL
		} literal;
		op_t op;
		int errno;
	} vals;
} tokenizer_t;

// init the tokenizer
void tokenizer_module_init(void);

// initialize a tokenizer 
tokenizer_t *tokenizer_create(const char *line, int cstyle_allowed);

// set to next token; return true when there is a valid token
// if allow_index is true, comma is folded into ",x" etc addressig mode tokens
// if allow_colon is true, comments may include a colon
bool_t tokenizer_next(tokenizer_t *tok, int allow_index);
bool_t tokenizer_next_prefix(tokenizer_t *tok);
bool_t tokenizer_next_comment(tokenizer_t *tok, int allow_colon);

// clear token, so _next will return the same token again
void tokenizer_rewind(tokenizer_t *tok);

void tokenizer_free(tokenizer_t *tok);

op_details_t *tokenizer_op_details(op_t op);

#endif


