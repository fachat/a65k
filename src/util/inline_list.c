/****************************************************************************

    list handling
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>

#include "types.h"
#include "inline_list.h"


static type_t ilist_memtype = {
	"ilist",
	sizeof(ilist_t)
};

ilist_t *inline_list_init(int bucket_size, const type_t *type, void (*init_item)(void*)) {

	ilist_t *ilist = mem_alloc(&ilist_memtype);

	ilist->cap = bucket_size;
	ilist->len = 0;
	ilist->type = type;
	ilist->init_item = init_item;
	ilist->values = mem_alloc_n(bucket_size, type);

	return ilist;
}


