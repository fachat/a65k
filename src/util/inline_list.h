/****************************************************************************

    list handling
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

/*
   A list that stores the entries in an array.
   (this is different from arraylist, that only stores pointers to the entry)

   Because this requires that a pointer to the memory for the object must
   be returned by the list, not added to it, the API changes and
   inline_list does not implement the general list interface.
*/


#ifndef INLINE_LIST_H
#define INLINE_LIST_H

#include <stdio.h>

#include "types.h"
#include "mem.h"


typedef struct {
	int		cap;		// capacity of list
	int		len;		// number of items in list
	int		popped;		// flag, if true add is not allowed anymore
	const type_t	*type;		// type info for array element
					// function to init a new item
	void 		(*init_item)(void*);
					// list of inlined items
	void		*values;
} ilist_t;

ilist_t *inline_list_init(int bucket_size, const type_t *type, void (*init_item)(void*));

static inline int ilist_len(const ilist_t *il) {
	return il->len;
}

static inline void* ilist_get_internal(const ilist_t *ilist, int itemno) {
	char *ptr = (char*) ilist->values;
	return (void*) &ptr[ilist->type->sizeoftype * itemno];
}

static inline void* ilist_add(ilist_t *ilist) {

	if (ilist->popped) {
		// we have popped stuff, so we cannot overwrite it!
		return NULL;
	}
	
	ilist->len++;
	if (ilist->len >= ilist->cap) {
		ilist->cap = ilist->cap * 2;
		ilist->values = mem_realloc_n(ilist->cap, ilist->type, ilist->values);
	}
	return ilist_get_internal(ilist, ilist->len - 1);
}

static inline void* ilist_pop(ilist_t *ilist) {

	if (ilist->len > 0) {
		ilist->popped = 1;
		ilist->len --;
		return ilist_get_internal(ilist, ilist->len);
	} else {
		// TODO: Error
		return NULL;
	}
}

static inline void* ilist_get(const ilist_t *ilist, int itemno) {

	if (ilist->len <= itemno) {
		return NULL;
	}
	return ilist_get_internal(ilist, itemno);
}

static inline void* ilist_last(const ilist_t *ilist) {

	if (ilist->len == 0) {
		return NULL;
	}
	return ilist_get_internal(ilist, ilist->len - 1);
}

#endif

