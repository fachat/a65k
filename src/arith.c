/****************************************************************************

    arithmetic operations and expressions
    Copyright (C) 2015,2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include "types.h"
#include "mem.h"
#include "tokenizer.h"
#include "arith.h"

typedef enum {
	EXP_VAL,
	EXP_OP,
	EXP_IND,
	EXP_END
} expect_t;

static err_t arith_parse_int(tokenizer_t *tok, int allow_index, const ilist_t **ext_anode, op_t closing) {

	err_t rv = E_OK;

	ilist_t *list = inline_list_init(10, &anode_memtype, (void(*)(void*))anode_init);

	anode_t *anode = ilist_add(list);

	// when true, next token should be a value
	expect_t expect = EXP_VAL;

	//op_t unary = OP_NONE;

	//while (tokenizer_next(tok, allow_index)) {
	do {
		//printf("loop at: type=%c len=%d op=%d -> %s (n=%p, subv.type='%c')\n", 
		//	tok->type, tok->len, tok->vals.op, tok->line+tok->ptr, anode, anode->val.subv.type);
		switch(expect) {
		case EXP_VAL:
			switch(tok->type) {
			case T_INIT:
				break;
			case T_BRACKET:
				anode->type = A_BRACKET;
				anode->val.subv.type = tok->vals.op;
				if (tokenizer_next(tok, allow_index)) {
					// NOTE cast necessary due to stupid standards nitpicking by gcc 
					// https://stackoverflow.com/questions/28701376/incompatible-pointer-types-and-constness
					arith_parse(tok, allow_index, (const ilist_t**) &anode->val.subv.value);

					if (tok->type != T_BRACKET) {
						rv = E_SYNTAX;
					}
				}
				expect = EXP_OP;
				anode = ilist_add(list);
				break;
			case T_NAME:
				// TODO
				break;
			case T_LITERAL:
				// new node for literal value
				anode->type = A_VALUE;
				anode->val.intv.type = tok->vals.literal.type;
				anode->val.intv.value = tok->vals.literal.value;
				expect = EXP_OP;
				anode = ilist_add(list);
				break;
			case T_TOKEN:
				// modifier?
				if (tokenizer_op_details(tok->vals.op)->is_modifier) {
					if (anode->modifier != AMOD_NONE) {
						// TODO log error
					} else {
						// conveniently mapped
						anode->modifier = tok->vals.op;
					}
					break;
				}
				// unary operator, like inversion, negative
/*
				if (unary != OP_NONE) {
					new_anode = anode_init(A_VALUE, anode);
					new_anode->val.intv.type = LIT_NONE;
					new_anode->op = unary;
					unary = OP_NONE;
					anode = new_anode;
					new_anode = NULL;
				}
				unary = tok->vals.op;
*/
				rv = E_SYNTAX;
				break;
			case T_STRING:
				if (tok->vals.string.len == 1) {
					anode->type = A_VALUE;
					anode->val.intv.type = LIT_CHAR;
					anode->val.intv.value = 0xff&tok->line[tok->vals.string.ptr];
					expect = EXP_OP;
					anode = ilist_add(list);
					break;
				} else
				if (tok->vals.string.len == 2) {
					anode->type = A_VALUE;
					anode->val.intv.type = LIT_TWOCHAR;
					// TODO: byte order
					anode->val.intv.value = (tok->line[tok->vals.string.ptr] << 8) + (tok->line[tok->vals.string.ptr+1]&0xff);
					expect = EXP_OP;
					anode = ilist_add(list);
					break;
				}
			case T_ERROR:
				rv = E_SYNTAX;
			case T_END:
				break;
			}
			break;
		case EXP_OP:	
			switch(tok->type) {
			case T_BRACKET:
				// closing bracket
				goto end;
				break;
			case T_TOKEN:
				if (tokenizer_op_details(tok->vals.op)->is_index) {
					anode->type = A_INDEX;
					anode->op = tok->vals.op;
					expect = EXP_IND;
					break;
				}
				if (tokenizer_op_details(tok->vals.op)->is_binary) {
					anode->op = tok->vals.op;
					expect = EXP_VAL;
					break;
				}
				goto end;
			case T_INIT:
			case T_NAME:
			case T_STRING:
			case T_LITERAL:
			case T_ERROR:
				rv = E_SYNTAX;
			case T_END:
				break;
			}
			break;
		case EXP_IND:
			if (tokenizer_op_details(tok->vals.op)->is_index) {
				anode = ilist_add(list);
				anode->type = A_INDEX;
				anode->op = tok->vals.op;
				expect = EXP_IND;
				break;
			}
			// fall-through
		case EXP_END:
end:
			if (closing) {
				if (tok->type == T_BRACKET) {
					if (tok->vals.op != closing) {
						rv = E_ARITH_CLOSING;
						goto exit;
					}
					
				}
			}
			goto exit;
			break;
		}	
	}
	while ((rv == E_OK) && tokenizer_next(tok, allow_index));
exit:
	if (anode->op == OP_NONE) {
		ilist_pop(list);
	}
	*ext_anode = list;

	//printf("return at tok: type=%c len=%d -> %s (n=%p, subv.type='%c')\n", 
	//	tok->type, tok->len, tok->line+tok->ptr, anode, anode->val.subv.type);

	return rv;
}


err_t arith_parse(tokenizer_t *tok, int allow_index, const ilist_t **ext_anode) {

	return arith_parse_int(tok, allow_index, ext_anode, OP_NONE);
}


