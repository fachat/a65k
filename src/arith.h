
/****************************************************************************

    arithmetic operations and expressions
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef ARITH_H
#define ARITH_H

#include "inline_list.h"
#include "err.h"

typedef signed long maxval_t;

/*
The anode_t struct is a node in an arithmetic expression
It is linked together in a tree-structure like such:

	(foo + 3) << 2, "foo", -1

	node0(A_BRACKET, AB_RND, OP_SHIFTLEFT), 
          |  |    
          |  |    
	  |  |
	  |  +-(value)->node1(A_LABEL, OP_ADD),
          |             nodw2(A_VALUE, AV_DEC, OP_NONE)
          |
        node3(A_VALUE, AV_DEC, OP_END)

	node4(A_VALUE, AB_STRD, OP_END)
	 
	node5(MOD_NEG, A_VALUE, AV_DEC)

Note: comma-separated values are only allowed in special
pseudo-opcodes and are handled as lists of values there.

The parser result also is a list. Only where brackets are
involved, there are sublists.
*/

// we might as well make our constants printable..
typedef enum {
	A_NONE		='-',	// placeholder
	A_BRACKET	='b',	// open bracket
	A_VALUE		='v',	// value
	A_STRING	='s',	// string
	A_LABEL		='l',	// label reference
	A_INDEX		='i',	// index, ",x"
} a_type;

typedef enum {
	// A_BRACKET subtypes (not necessarily all are used)
	AB_NONE		= 0,
	AB_RND		='(',		// (
	AB_ANG		='<',		// <
	AB_RCT		='[',		// [
	AB_CRV		='{',		// {
	AB_DBLRND	='d',		// ((
	AB_DBLRCT	='D',		// [[
} asub_type;

typedef enum {
	AMOD_NONE 	=0,		// none
	AMOD_LOW	=OP_LESS,	// low byte selector (bits 0-7)
	AMOD_HIGH	=OP_LARGER,	// high byte selector (bits 8-15)
} amod_type;

/*
 The anode struct allows building the AST for an arithmetic expression.
*/
typedef struct anode_s anode_t;

struct anode_s {
	// type, incl. brackets, arithm. ops etc
	a_type		type;
	// modifier
	amod_type	modifier;
	// operation
	op_t		op;
	// value (depending on a_type)
	union {
	  // A_VALUE
	  struct {
	    maxval_t 	value;
	    littype_t	type;	// note: maybe LIT_NONE in case of stacked unaries
	  } intv;
	  // A_STRING
	  struct {
	    const char 	*value;
	    int 	len;
	    quotetype_t	type;
	  } strv;
	  // A_LABEL
	  // TODO
	  // A_BRACKET
	  struct {
	    ilist_t *value;
	    asub_type	type;
	  } subv;
	} val;
};


static type_t anode_memtype = {
	"label",
	sizeof(anode_t)
};

static inline anode_t *anode_init() {
	anode_t *anode = mem_alloc(&anode_memtype);
	
	anode->type = A_NONE;
	anode->op = OP_NONE;
	anode->modifier = AMOD_NONE;

	return anode;
}

err_t arith_parse(tokenizer_t *tok, int allow_index, const ilist_t **anode);

#endif

