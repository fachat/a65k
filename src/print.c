/****************************************************************************

    print output
    Copyright (C) 2012,2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdarg.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>

#include "print-config.h"
#include "print.h"
#include "arith.h"

// filler spaces
static const char spaces[] = "                                        ";

// column widths for fixed width output, index corresponds to printer_col_t
static const int widths[] = { 
	6, 	// PRT_LINENO
	5,	// PRT_ADDR
	9,	// PRT_DATA
	11,	// PRT_LABEL
	4,	// PRT_OPERATION
	20,	// PRT_PARAM
	30	// PRT_COMMENT
};

void print_module_init() {

	print_config_init();
}

static type_t printer_memtype = {
        "printer_t",
        sizeof(printer_t)
};

// initialize printer output channel
printer_t *print_init() {

	print_config_t *cfg = print_current_config();

	const char *filename = cfg->filename;
	FILE *f = NULL;
	if (filename != NULL && strlen(filename) > 0) {
		f = fopen(filename, "w");
		if (f == NULL) {
			log_error("Could not open output file '%s' -> %s", filename, strerror(errno));
			return NULL;
		}
	}

	printer_t *prt = mem_alloc(&printer_memtype);

	prt->outfile = f;
	prt->cfg = cfg;
	prt->current_col = PRT_LINENO;

	int p = 0;
	for (int i = PRT_LINENO; i < PRT_END; i++) {
		prt->col_width[i] = p;
		p += widths[i];
	}

	print_clr(prt);

	prt->was_empty = 0;

	return prt;
}

static int print_getlen(printer_t *prt) {
	return prt->buflen;
}

static void print_setcol(printer_t *prt, int col) {
	
	int buflen = print_getlen(prt);

	int todo = col - buflen;
	int slen = strlen(spaces);

	if (todo <= 0) {
		todo = 1;
	}

	while (todo > 0) {
		int step = slen - todo;
		if (step < 0) {
			step = 0;
		}
		print(prt, prt->current_col, "%s", spaces + step);
		todo -= slen + step;
	}
}


void print(printer_t *prt, print_col_t col_no, const char *pattern, ...) {

	if (col_no != prt->current_col) {
		// target fixed width column
		if (col_no == PRT_COMMENT && prt->current_col == PRT_LINENO) {
			// shift comment to left if only comment on line
			col_no = PRT_LABEL;
		}
		int trg = prt->col_width[col_no];
		print_setcol(prt, trg);
		prt->current_col = col_no;
	}

        va_list ap;
        va_start(ap, pattern);

	int buflen = print_getlen(prt);

        int r = vsnprintf(prt->buf + buflen, BUF_LEN - buflen, pattern, ap);
        if (r < 0 || r > BUF_LEN) {
                // error
                log_error("Error printing %d\n", r);
                return;
        }
	prt->buflen += r;
}

void print_out(printer_t *prt) {

	if (prt->buflen == 0) {
		if (prt->cfg->collapse_empty && prt->was_empty) {
			return;
		}
		prt->was_empty = 1;
	} else {
		prt->was_empty = 0;
	}

	if (prt->outfile == NULL) {
        	printf("%s\n", prt->buf);
	} else {
        	fprintf(prt->outfile, "%s\n", prt->buf);
	}
	print_clr(prt);
}

void print_clr(printer_t *prt) {
	prt->buf[0] = 0;
	prt->buflen = 0;
	prt->current_col = PRT_LINENO;
}

void print_close(printer_t *prt) {

	if (prt->outfile) {
		fclose(prt->outfile);
	}
}


