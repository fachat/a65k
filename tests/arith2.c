

#include <stdio.h>

#include "inline_list.h"
#include "tokenizer.h"
#include "print.h"
#include "arith.h"


void print(const char *val, int allow_index) {

	do_print("---> %s%s", val, allow_index ? " (with index)":"");

	tokenizer_t *tok = tokenizer_create(val, 0);

	const ilist_t *out = NULL;

	arith_parse(tok, allow_index, &out);

	print_debug_arith(out);
	
	do_print("   > next is type %c, %s", tok->type, tok->line + tok->ptr);
}

int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

	tokenizer_module_init();

	print("123,x", 1);
	print("$123,X", 0);
	print("$123,X", 1);
	print("($123),y", 0);
	print("($123),y", 1);
	print("($123,X)", 0);
	print("($123,X)", 1);
	print("($123,s,X)", 1);

}



