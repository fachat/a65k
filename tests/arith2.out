WRN:---> 123,x (with index)
WRN:  A:len=2
WRN:    type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:    type=i, modifier=0 ( ), op=148( )
WRN:   > next is type x, 
WRN:---> $123,X
WRN:  A:len=1
WRN:    type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:   > next is type t, ,X
WRN:---> $123,X (with index)
WRN:  A:len=2
WRN:    type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:    type=i, modifier=0 ( ), op=148( )
WRN:   > next is type x, 
WRN:---> ($123),y
WRN:  A:len=1
WRN:    type=b, modifier=0( ), op=0( ) btype=40 (()
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:   > next is type t, ,y
WRN:---> ($123),y (with index)
WRN:  A:len=2
WRN:    type=b, modifier=0( ), op=0( ) btype=40 (()
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:    type=i, modifier=0 ( ), op=149( )
WRN:   > next is type x, 
WRN:---> ($123,X)
WRN:  A:len=1
WRN:    type=b, modifier=0( ), op=0( ) btype=40 (()
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:   > next is type b, )
WRN:---> ($123,X) (with index)
WRN:  A:len=1
WRN:    type=b, modifier=0( ), op=0( ) btype=40 (()
WRN:    A:len=2
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:      type=i, modifier=0 ( ), op=148( )
WRN:   > next is type x, 
WRN:---> ($123,s,X) (with index)
WRN:  A:len=1
WRN:    type=b, modifier=0( ), op=0( ) btype=40 (()
WRN:    A:len=3
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)291/123
WRN:      type=i, modifier=0 ( ), op=151( )
WRN:      type=i, modifier=0 ( ), op=148( )
WRN:   > next is type x, 
